tests/end-to-end.rs

```diff
@@ -22,86 +41,366 @@ macro_rules! test_case {
-            let actual = {
-                let mut output = Vec::with_capacity(expected.len());
-                let cfn: CloudformationParseTree = serde_yaml::from_str(include_str!(concat!(
-                    "end-to-end/",
-                    stringify!($name),
-                    "/template.yml"
-                )))
-                .unwrap();
+            // Check that the original cloudformation template is valid
+            println!("Verifying a CloudFormation stack can be created from original template");
+            let original_template =
+                include_str!(concat!("end-to-end/", stringify!($name), "/template.json"));
+            create_stack($stack_name, original_template);
+
+            println!("Creating the cdk stack definition");
+            let cdk_stack_definition = {
+                let mut output = Vec::new();
+                let cfn: CloudformationParseTree = serde_yaml::from_str(original_template).unwrap();
                 let ir = CloudformationProgramIr::from(cfn).unwrap();
                 ir.synthesize($synthesizer, &mut output, $stack_name)
                     .unwrap();
                 String::from_utf8(output).unwrap()
             };
 
-            let _update_snapshots = UpdateSnapshot::new(
-                concat!("end-to-end/", stringify!($name), "/", $expected),
-                &actual,
-                &expected,
+            let mut snapshots_zip =
+                get_zip_archive_from_bytes(include_bytes!("./end-to-end-test-snapshots.zip"));
+
+            println!("Checking for cdk stack definition in the expected output");
+            let expected_outputs_dir = format!("{}/{}", stringify!($name), stringify!($lang));
+            let expected_cdk_stack_def_filename =
+                format!("{expected_outputs_dir}/{}", $cdk_stack_filename);
+            check_cdk_stack_def_matches_expected(
+                &cdk_stack_definition,
+                &expected_cdk_stack_def_filename,
+                &mut snapshots_zip,
             );
-            assert_eq!(expected, actual);
+
+            let test_working_dir = format!(
+                "tests/end-to-end/{}-{}-working-dir",
+                stringify!($name),
+                stringify!($lang)
+            );
+            let language_boilerplate_dir = format!(
+                "tests/end-to-end/app-boilerplate-files/{}",
+                stringify!($lang)
+            );
+            synth_cdk_app(
+                &cdk_stack_definition,
+                $stack_name,
+                $cdk_stack_filename,
+                $cdk_app_filename,
+                &language_boilerplate_dir,
+                &test_working_dir,
+                &expected_outputs_dir,
+                &mut snapshots_zip,
+            );
+
+            diff_original_template_with_new_templates(stringify!($name), &test_working_dir);
+
+            // update snapshots
+            update_snapshots(
```

Comment:
options type for this? If it's only ever called in one place this is fine, but if we have to call it more than once the options type would make this easier